torch/amp/grad_scaling.py

```diff
@@ -0,0 +1,382 @@
+import torch
+import types
+from torch._six import container_abcs
+
+
+_default_init_scale = 2.**24
+_default_scale_growth_factor = 1.001
+_default_scale_backoff_factor = 0.5
+
+
+def _recommend_init_scale():
+    return torch.full((1,), _default_init_scale, dtype=torch.float32, device="cuda")
+
+
+def scale_outputs(outputs, current_scale, scaling_enabled=True):
+    r"""
+    Scales a network's output(s) by ``current_scale``, so that when ``backward()`` is called on the scaled outputs,
+    the resulting gradients will also be scaled by ``current_scale``.
+
+    Arguments:
+        outputs (Tensor or iteratble of Tensors):  Outputs to multiply by ``current_scale``.
+        current_scale (float or torch.cuda.FloatTensor):  Scale factor by which to multiply the outputs.
+        scaling_enabled (bool, default=True):  If False, :func:`scale_outputs` is a no-op, and returns ``outputs``
+        directly.
+
+    Returns:
+        Outputs scaled by ``current_scale``, or ``outputs`` if ``scaling_enabled=False``.
+    """
+    if not scaling_enabled:
+        return outputs
+
+    # Not necessary.
+    # if isinstance(current_scale, float):
+    #     current_scale = torch.full((1,), current_scale, dtype=torch.float32, device="cuda")
+
+    def scale(val):
+        if isinstance(val, torch.Tensor):
+            assert val.is_cuda
+            return val * current_scale
+        elif isinstance(val, container_abcs.Iterable):
+            return type(val)(scale(v) for v in val)
+        else:
+            raise ValueError("outputs must be a Tensor or an interable of Tensors")
+
+    return scale(outputs)
+
+
+def unscale_and_step(self,
+                     closure=None,
+                     current_scale=None,
+                     scale_growth_factor=_default_scale_growth_factor,
+                     scale_backoff_factor=_default_scale_backoff_factor,
+                     scale_scheduler=None,
+                     skip_if_inf=True,
+                     found_inf=None,
+                     scaling_enabled=True):
+    r"""
+    :func:`unscale_and_step` is patched onto an optimizer instance by :func:`add_amp_attributes`.  Since
```

Comment:
There's probably a better way to hide this function and expose it only after you've attached it to an optimizer.